---
title: "Investigating the sequencing_summary file of barcode 4 (phenol-chloroform) from SUPER accuracy rebasecalling barcodes trimmed)"
output: html_notebook
---
  
Tidyverse package loaded at the start of this notebook, as it is a convenient package that includes tidyr, ggplot2, dplyramong other useful packages.

```{r setup}
#install.packages("tidyverse")
library(tidyverse)
library(scales) #for the percentage function below
```


**Load in the summary sequencing file for the extraction method/barcode being examined**

```{r}
sequencing_summary <- read_delim ("C:/Users/keegk/OneDrive - moredungroup/Karen Keegan/Nanopore/SQKRBK004/Extraction_experiment_run_22.2.23/Rebasecalling/Rebasecalled_trimmed_super_accuracy/barcode_4/sequencing_summary_barcode4.txt")


extexp2025_barcode4_remove_reads <- sequencing_summary %>%
                      filter(!barcode_arrangement == "barcode04") %>%
                      select(read_id) #this is isolating the reads that are on the wrong barcode, so I can use this read Id list to remove these reads in the blast text files (blast text files are assigned taxonomy in the notebook "taxonomizr_blast_tidied_2025.Rmd")

extexp2025_barcode4_remove_reads <- save(extexp2025_barcode4_remove_reads, file = "extexp2025_barcode4_remove_reads")
load("extexp2025_barcode4_remove_reads") #this was saved with 2025 written after it but the actual file name loads in without the 2025 bit after it just FYI.


sequencing_summary <- sequencing_summary %>%
                      filter(barcode_arrangement == "barcode04") 
```


**Read QC from the summary sequencing file**
  
```{r sequence lengths}

#Total number of reads
no_reads <- nrow(sequencing_summary) # total number of reads
No_reads <- formatC(no_reads, format = "fg", digits = 2, big.mark = ",")

# percentage of reads that passed q score:
passed_reads <- subset(sequencing_summary, passes_filtering == "TRUE")
passed <- nrow(passed_reads)
percentage <- passed/no_reads# this is my passed reads over my total reads
#Percentage <- format(round(percentage, 2)) # percentage of reads that passed
Percentage <- label_percent()(percentage)
print(paste(Percentage, "is the percentage of reads that passed"))

mean_read <- mean(sequencing_summary$sequence_length_template)#average read length
Mean_read = format(round(mean_read, 1))
Mean_read <- formatC(mean_read, format = "fg", digits = 2, big.mark = ",")

max_read <- max(sequencing_summary$sequence_length_template) #highest read length
Max_read <- formatC(max_read, format = "fg", digits = 2, big.mark = ",")

median <- median(sequencing_summary$sequence_length_template) #median (N50 read no.)
Median = formatC(median, format = "fg", digits = 2, big.mark = ",")

no_bases <- sum(sequencing_summary$sequence_length_template)#total number of bases per extraction method
No_bases <- formatC(no_bases, format = "fg", digits = 2, big.mark = ",")

mean_q_score <- mean(sequencing_summary$mean_qscore_template) #mean q score
Mean_q_score = formatC(mean_q_score, format = "fg", digits = 2, big.mark = ",")

max_read_q_score <- max(passed_reads$sequence_length_template) 
                    

#results printed out:

print(paste(No_reads, "is the no. of reads that passed"))
print(paste(Percentage, "is the percentage of reads that passed"))
print(paste(Mean_read, "is the mean read"))
print(paste(Max_read, "is the max read"))
print(paste(max_read_q_score, "is the max read passing Q score"))
print(paste(Mean_q_score, "is the mean Q score"))
print(paste(Median, "is the median read"))
print(paste(No_bases, "is the no. of bases"))


####beginning histogram###

nbins <- floor(sqrt(nrow(sequencing_summary)))

sequencing_summary %>%
  ggplot(aes(x = sequence_length_template)) +
  geom_histogram (bins =nbins, fill = 'blue') +
#  geom_text(aes(347000, 1500, label=paste("Total number of reads = ", No_reads))) +
 # geom_text(aes(347000, 1400, label=paste("Total number of bases = ", No_bases))) +
 # geom_text(aes(347000, 1300, label=paste("Mean read length = ", Mean_read))) +
 # geom_text(aes(347000, 1200, label=paste("Max read length = ", Max_read))) +
  #geom_text(aes(347000, 1100, label=paste("N50 = ", Median))) +
  #geom_text(aes(347000, 1000, label=paste("Mean Q score = ", Mean_q_score))) +
 # geom_text(aes(347000, 900, label=paste("Percentage of reads that passed = ", Percentage))) +
      labs(x = "read length (bp)", y = "count")#, title = "Phenol-chloroform  read length distribution")

sequencing_summary %>%
  ggplot(aes(x=sequence_length_template)) +
  geom_histogram(bins=nbins, fill = 'blue') +
  coord_cartesian(ylim = c(0, 100)) +
  labs(x = "Read length (kb)", y = "count")#,
     #title = "Phenol-chloroform read length distribution")


```

**Histogram of Q score distribution across the reads**

```{r}
sequencing_summary%>% 
  ggplot(aes(x = mean_qscore_template)) +
  geom_histogram(bins = nbins) +
  labs(x = "read mean quality score", y = "count",
       title = "read quality score historgram")

mean(sequencing_summary$mean_qscore_template)
```
