---
title: "Adding extraction method column and merging blast text files 2025"
output: html_notebook
---

```{r setup}
library(tidyverse)
library(stringr)
```


Barcode 1:

```{r}
 
barcode_1.txt <- read.table("parsed_extexp2025_bar1_blast.txt", sep=",", header = TRUE)

barcode_1.1 <- barcode_1.txt %>%
   separate_rows(subject.sci.names, subject.tax.ids, sep = ";")

grep(";", barcode_1.1$subject.sci.names) 

extraction.method <- c("FAST DNA")
barcode_1.1$extraction.method <- extraction.method


barcode_1.1$subject.tax.ids <- as.character(barcode_1.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.

#from the summary sequencing file QC analysis (notebook "<extraction method>_summary_seq.Rmd - there is a unique notebook for each extraction method), I noticed that there were incorrect barcodes (not many across all the extraction methods/barcode summary seq files) mixed in with the barcode of interest. I filtered these out when doing the read QC (no. of reads, max read length etc), but I should also insure that they are being removed in the taxonomy stage too. Now the blast text files in this notebook do not have info on the barcode - i add this info in myself, so theres no way for me to scan the blast text file to find reads assigned to a barcode that shouldn't be there. Instead im going to screen the blast dataframe for read ID's of the incorrect barcodes in the summary seq files and essentially remove these read ID's from the blast text files.

#i have written code in the <extraction method>_summary_seq.Rmd that has extracted the read ID's we want to remove per barcode, so I am just loading these in here but you can go back into the original notebook to see the full code for this.

load("extexp2025_barcode1_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode1_remove_reads <- rename(extexp2025_barcode1_remove_reads, query.acc = read_id)

barcode_1.2 <- barcode_1.1 %>% anti_join(extexp2025_barcode1_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcode1_remove_reads from barcode_1.1 

#testing whether the anti join code works by grabbing read ids that ARE in the blast text file, running antijoin code and seeing if they are removed:

known_reads <- barcode_1.2 %>% #manually looking at dataframe and grabbing first two reads
              filter(query.acc %in% c("efb2f860-1b1c-4afd-ad00-5803a75ad0fa", "fba2901a-3500-4361-8e46-ef1eadbdbf5e"))
barcode_1_removal_test <- barcode_1.1 %>% anti_join(known_reads) #confirms antijoin code works, as this dataset has 5 reads removed, which correspond to the number of reads in known_reads dataset (fb2f860-1b1c-4afd-ad00-5803a75ad0fa has one alignment, fba2901a-3500-4361-8e46-ef1eadbdbf5e has 4 alignments)
```

Barcode 2

```{r}
 
barcode_2.txt <- read.table("parsed_extexp2025_bar2_blast.txt", sep=",",  header = TRUE)

barcode_2.1 <- barcode_2.txt %>% separate_rows(subject.sci.names, subject.tax.ids, sep = ";")
grep(";", barcode_2.1$subject.sci.names) 

extraction.method <- c("MagMAX")
barcode_2.1$extraction.method <- extraction.method


barcode_2.1$subject.tax.ids <- as.character(barcode_2.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.

load("extexp2025_barcode2_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode2_remove_reads <- rename(extexp2025_barcode2_remove_reads, query.acc = read_id)

barcode_2.2 <- barcode_2.1 %>% anti_join(extexp2025_barcode2_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcodeX_remove_reads from barcode_X.X 

```

Barcode 3

```{r}
   
barcode_3.txt <- read.table("parsed_extexp2025_bar3_blast.txt", sep=",",  header = TRUE)

barcode_3.1 <- barcode_3.txt %>% separate_rows(subject.sci.names, subject.tax.ids, sep = ";")
grep(";", barcode_3.1$subject.sci.names) 

extraction.method <- c("Three peaks")
barcode_3.1$extraction.method <- extraction.method


barcode_3.1$subject.tax.ids <- as.character(barcode_3.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.

load("extexp2025_barcode3_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode3_remove_reads <- rename(extexp2025_barcode3_remove_reads, query.acc = read_id)

barcode_3.2 <- barcode_3.1 %>% anti_join(extexp2025_barcode3_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcodeX_remove_reads from barcode_X.X 

```

Barcode 4

```{r}

barcode_4.txt <- read.table("parsed_extexp2025_bar4_blast.txt", sep=",",  header = TRUE)


barcode_4.1 <- barcode_4.txt %>% separate_rows(subject.sci.names, subject.tax.ids, sep = ";")
grep(";", barcode_4.1$subject.sci.names) 

extraction.method <- c("PC")
barcode_4.1$extraction.method <- extraction.method


barcode_4.1$subject.tax.ids <- as.character(barcode_4.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.


load("extexp2025_barcode4_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode4_remove_reads <- rename(extexp2025_barcode4_remove_reads, query.acc = read_id)

barcode_4.2 <- barcode_4.1 %>% anti_join(extexp2025_barcode4_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcodeX_remove_reads from barcode_X.X 
```

Barcode 5

```{r}

barcode_5.txt <- read.table("parsed_extexp2025_bar5_blast.txt", sep=",",  header = TRUE)

barcode_5.1 <- barcode_5.txt %>% separate_rows(subject.sci.names, subject.tax.ids, sep = ";")
grep(";", barcode_5.1$subject.sci.names) 

extraction.method <- c("UCP")
barcode_5.1$extraction.method <- extraction.method


barcode_5.1$subject.tax.ids <- as.character(barcode_5.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.

load("extexp2025_barcode5_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode5_remove_reads <- rename(extexp2025_barcode5_remove_reads, query.acc = read_id)

barcode_5.2 <- barcode_5.1 %>% anti_join(extexp2025_barcode5_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcodeX_remove_reads from barcode_X.X 
```

Barcode 6

```{r}

barcode_6.txt <- read.table("parsed_extexp2025_bar6_blast.txt", sep=",",  header = TRUE)

barcode_6.1 <- barcode_6.txt %>% separate_rows(subject.sci.names, subject.tax.ids, sep = ";")
grep(";", barcode_6.1$subject.sci.names) 

extraction.method <- c("MagAttract")
barcode_6.1$extraction.method <- extraction.method


barcode_6.1$subject.tax.ids <- as.character(barcode_6.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.


load("extexp2025_barcode6_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode6_remove_reads <- rename(extexp2025_barcode6_remove_reads, query.acc = read_id)

barcode_6.2 <- barcode_6.1 %>% anti_join(extexp2025_barcode6_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcodeX_remove_reads from barcode_X.X 
```

Barcode 7

```{r}

barcode_7.txt <- read.table("parsed_extexp2025_bar7_blast.txt", sep=",",  header = TRUE)

barcode_7.1 <- barcode_7.txt %>% separate_rows(subject.sci.names, subject.tax.ids, sep = ";")
grep(";", barcode_7.1$subject.sci.names) 

extraction.method <- c("Genomic-tip")
barcode_7.1$extraction.method <- extraction.method

barcode_7.1$subject.tax.ids <- as.character(barcode_7.1$subject.tax.ids) # for some reason the bind_rows function below is throwing up an error that some of the subject.tax.id columns are chracters and others are integers across all the barcode text files, so setting every barcodes column to character before merging.


load("extexp2025_barcode7_remove_reads") #this is a list of read ID's taken from the summary seq file for barcode 1, that were from incorrect barcoodes (barcode/tag switching potentially). These may have already been removed based on the filtering criteria in BLAST, but for completeness, below is code to remove these read ID's from the blast text file for this barcode:

#need to change the name of the column in extexp2025_barcode1_remove_reads from read_id to query.acc (they are the same thing) so we can antijoin these two dataframes using query.acc as the linker column:

extexp2025_barcode7_remove_reads <- rename(extexp2025_barcode7_remove_reads, query.acc = read_id)

barcode_7.2 <- barcode_7.1 %>% anti_join(extexp2025_barcode7_remove_reads) #this says, join these two dataframes, but remove any query.acc variables in extexp2025_barcodeX_remove_reads from barcode_X.X 

```
No merging all dataframes using tidyverse bind_rows

Just to note, that the code to remove wrong barcodes from each barcodes blast text file resulted in no wrong barcodes being removed (the code was tested in barcode 1 chunk above so the antijoin code used does work, implying that the wrong barcodes were filtered out at the blast stage. Why might this be? Reminder that the blast read filtering thus far is 90% id, evalue less than 1 x 10^-12 and max target seqs of 10. We have yet to add the further filtering (loose/stringent) to these blast hits. So potentially the wrong barcodes are just noisey and the hits to a subject do not pass the percent id or e value test - the only way i can think to test this is reblasting everything without the filters, but this is quite a lot of work, when we have already illustrated that these reads are not carrying through to our final blast results. 
```{r}
combined_blast <- bind_rows(barcode_1.2, barcode_2.2, barcode_3.2,barcode_4.2,barcode_5.2,barcode_6.2,barcode_7.2)
View(combined_blast)
#nrow(combined_blast) #42965
nrow(combined_blast)
#combined_blast_distinct<- dplyr::distinct(combined_blast, .keep_all = TRUE)
#nrow(combined_blast_distinct)
#View(combined_blast_distinct)
#combined_blast[11434, ]# ok so both of these row numbers have two taxi ids assigned to them. This must be an inherent error in ncbi assignment giving the same species multiple tax ids? to correct this, maybe i use str_split to just take the first entry?
#combined_blast$subject.tax.ids<- str_extract(combined_blast$subject.tax.ids, "\\d+") # the "\\d+" is a regex (regular expression) that means: \d a shorthand character class, which matches all numbers; it is the same as [0-9] and + is plus one or more of the expression. If i had this in brackets "(\\d+))" it would return all numeric values present but without the brackets, just returns one value

#combined_blast[11434, ]
```

Setting up a table with taxa info for the combined

```{r}
#install.packages("taxonomizr")
library(taxonomizr)
prepareDatabase(getAccessions = FALSE) # name is "nameNode.sqlite"
taxaID <- as.vector(combined_blast$subject.tax.ids)
#print(taxaID)
#taxa_id <- data.frame(combined_blast$subject.tax.ids)
#taxa_id <- as.data.frame(getTaxonomy(taxaID, sqlFile = "nameNode.sqlite",  desiredTaxa = c("superkingdom", "phylum", "genus", "species")))
taxa<-getTaxonomy(taxaID,sqlFile = "nameNode.sqlite", sep ) 
#print(taxa)

```

 converting row names to column using rownames_to_column:
```{r}
taxa <- data.frame(taxa)
taxa <- save(taxa, file = "taxa")
load("taxa") # when we use add taxonomy, the dataframe of this has subject tax ids from our blast file as a column (properly turned into a column with correct name in the line below this) and all the names only (not taxa ids) of each phylogenetic class associated with the subject tax id
taxa1<- tibble::rownames_to_column(taxa, var = "subject.tax.ids")
taxa1
```

We need to remove the decimal points and NA values from the column in taxa

```{r}
taxa1$subject.tax.ids <- str_extract(taxa1$subject.tax.ids, "\\d+")
taxa1
nrow(taxa1) 
View(taxa1)

#to note, i quickly cross referenced a handful of these subject tax ids by looking them up on ncbi taxonomy databse to confirm that the matches of the class names to a subject tax id are correct. All matched.
```

now I can merge these two tables using the column in common, subject.tax.ids. first i need to make the column values identical by sorting both columns from each table sequentially from low to high
```{r}

combined_blast_sorted <- combined_blast[order(combined_blast$subject.tax.ids), ] #order the combined blast results (every barcode) by subject tax ID
nrow(combined_blast_sorted) #COUNT THE NUMBER OF ROWS IN THIS dataset
#combined_blast_distinct <- combined_blast_sorted #filter this combined blast dataframe to only retain one hit per read
#nrow(combined_blast_distinct)
taxa_table_sorted <- taxa1[order(taxa1$subject.tax.ids), ] #order the taxonomy table by subject tax ids
taxa_table_distinct <- distinct(taxa_table_sorted) #filter this so i only have one subject tax id entry (rather than 10 of the same subject tax id). This is to reduce the size of this dataset, so that merges succesfully in the next R chunk
#View(taxa_table_sorted) 
```

And finally merging both taxonomy table and blast table using left join
```{r}
merged_with_taxa_2025 <- full_join(x = combined_blast_sorted, y = taxa_table_distinct, by ="subject.tax.ids") 
#combined_blast[2644, ]
#taxa1[2644, ]
#taxa_merge <- combined_blast_sorted %>% group_by(subject.tax.ids) %>% select(subject.tax.ids) %>% left_join(taxa_table_distinct, by ="subject.tax.ids") %>% na.omit()

#nrow(merged_with_taxa)
#merged_with_taxa_distinct_2025 <- distinct(merged_with_taxa_2025) # this has the exact same number of entries (n = 2290116) as merged_with_taxa. This would filter only out rows if you had any row that was identical in every single column.it doesnt necessarily filter out the exact same read if the other coulmns (such as alignment length) differ slightly between these different read hits. This is somewhat redundant code so is being excluded , as later on in the "loose_stringent_fuiltering" dataset, the filtering on multiple instances of the same read are dealt with.
#nrow(merged_with_taxa_distinct)
#view(merged_with_taxa)

merged_with_taxa_2025_distinct <- distinct(merged_with_taxa_2025)
merged_with_taxa_2025_distinct <- save(merged_with_taxa_2025_distinct, file = "merged_with_taxa_2025_distinct")
load("merged_with_taxa_2025_distinct") #this was saved with 2025 written after it but the actual file name loads in without the 2025 bit after it just FYI.

```

